### PR TITLE
webgpu: Support any component buffer

### PR DESCRIPTION
**Pull request by qjia7**
To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

- - -
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7426)

### CODE REVIEW

File: tfjs-backend-webgpu/src/conv2d_mm_webgpu.ts [hunk 1]
```diff
@@ -159,7 +159,7 @@ export class Conv2DMMProgram implements WebGPUProgram {
   dispatchLayout: {x: number[], y: number[], z: number[]};
   dispatch: [number, number, number];
   variableNames = ['x', 'W'];
-  variableTypes: string[];
+  variableComponents: number[];
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/conv2d_mm_webgpu.ts:162]: In future, can we change this to inputNames and inputComponents?

File: tfjs-backend-webgpu/src/conv2d_mm_webgpu.ts [hunk 2]
```diff
@@ -176,6 +176,7 @@ export class Conv2DMMProgram implements WebGPUProgram {
   tileInner: number;
   innerElementSize: number;
   isVec4?: boolean;
+  outputComponent = 1;
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/conv2d_mm_webgpu.ts:179]: We don't have to init it as 1 here

File: tfjs-backend-webgpu/src/conv_backprop_mm_webgpu.ts [hunk 3]
```diff
@@ -143,7 +144,8 @@ export class Conv2DDerInputMMProgram implements WebGPUProgram {
         this.elementsPerThread);
 
     if (this.isVec4) {
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/conv_backprop_mm_webgpu.ts:146]: We may need to clean up isVec4 here.

File: tfjs-backend-webgpu/src/webgpu_program.ts [hunk 4]
```diff
@@ -262,18 +278,19 @@ function makeShader(
...
       inputInfo
           .map(
               (x, i) => getInputSnippet(
                   x, outputData.shape,
-                  program.variableTypes ?
-                      (program.variableTypes[i] === 'vec4<f32>') :
-                      program.isVec4,
+                  program.variableComponents ?
+                      program.variableComponents[i] :
+                      program.outputComponent ? program.outputComponent : 1,
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/webgpu_program.ts:293]: Can we just pass program.outputComponent here?

**Review by gyagp — COMMENTED** [review]
In future PRs, we may need to 1) clean up isVec4 and everything is component 2) change variable to input

**Review by gyagp — COMMENTED** [review]
BTW, I didn't finish all the detailed reviews so I may have some other comments after you address the comments above.

File: tfjs-backend-webgpu/src/webgpu_program.ts [hunk 5]
```diff
@@ -158,6 +174,8 @@ function makeShader(
   const prefixSnippets: string[] = [];
   const flatWorkgroupSize = program.workgroupSize[0] *
       program.workgroupSize[1] * program.workgroupSize[2];
+  program.outputComponent =
+      program.outputComponent ? program.outputComponent : 1;
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/webgpu_program.ts:178]: It's not good to set a class member outside. We can leave this for future refactoring.

File: tfjs-backend-webgpu/src/webgpu_program.ts [hunk 6]
```diff
@@ -67,6 +68,21 @@ export const compileProgram =
       return pipeline;
     };
 
+export const typeSnippet = (component: number, type = 'f32') => {
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/webgpu_program.ts:71]: The parameter orders of typeSnippet and dataTypeToGPUType are different. We may also refactor this part in future.

**Review by gyagp — APPROVED** [review]
LGTM and we may do more refactorings in future.

**Review by xhcao — APPROVED** [review]
LGTM, thanks.

**Pull request merged**